src/avcodec/bitstream.rs

```diff
@@ -0,0 +1,134 @@
+// FIXME: I hae a feeling this (UnsafeDerefMut) shouldn't be needed, but settable! wants it
+use crate::shared::UnsafeDerefMut;
+use std::{ffi::CStr, ptr};
+
+// See https://blogs.gentoo.org/lu_zero/2016/03/21/bitstream-filtering/
+
+use crate::{
+    error::{Result, RsmpegError},
+    ffi,
+    shared::PointerUpgrade,
+};
+
+wrap_ref!(AVBitStreamFilter: ffi::AVBitStreamFilter);
+
+impl AVBitStreamFilter {
+    /// Find a bitstream filter instance with it's short name.
+    ///
+    /// See [`ffi::av_bsf_get_by_name`] for more info.
+    pub fn find_by_name(name: &CStr) -> Option<AVBitStreamFilterRef> {
+        unsafe { ffi::av_bsf_get_by_name(name.as_ptr()) }
+            .upgrade()
+            .map(|x| unsafe { AVBitStreamFilterRef::from_raw(x) })
+    }
+
+    // TODO: implement ffi::av_bsf_next
+    // const AVBitStreamFilter *av_bsf_next(void **opaque);
+}
+
+wrap!(AVBSFContext: ffi::AVBSFContext);
+settable!(AVBSFContext {
+    time_base_in: ffi::AVRational
+});
+
+impl AVBSFContext {
+    /// Create a new [`AVBSFContext`] instance, allocate private data and
+    /// initialize defaults for the given [`AVBitStreamFilterRef`].
+    ///
+    /// See [`ffi::av_bsf_alloc`] for more info.
+    pub fn new(filter: &ffi::AVBitStreamFilter) -> Self {
+        let mut bsfc_raw = ptr::null_mut();
+
+        unsafe {
+            ffi::av_bsf_alloc(filter, &mut bsfc_raw);
+            AVBSFContext::from_raw(bsfc_raw.upgrade().unwrap())
+        }
+    }
+    /// You need to initialize the context before you can send/receive_packets but after you set input parameters via [`AVBSFContext::set_par_in`].
+    ///
+    /// See [`ffi::av_bsf_init`] for more info.
+    pub fn init(&mut self) {
+        unsafe {
+            // TODO: Error checking
+            ffi::av_bsf_init(self.as_mut_ptr());
+        }
+    }
+    // OPTIMIZE: We should probably just be able to .upgrade() the returned c_int
+    /// Provide input data for the bitstream filter to process. To signal the end of the stream, send an NULL packet to the filter.
+    ///
+    /// See [`ffi::av_bsf_send_packet`] for more info.
+    pub fn send_packet(&mut self, packet: &mut ffi::AVPacket) -> Result<()> {
+        // TODO: Ensure init is called first
```

Comment:
I guess `av_bsf_send_packet` will emits error when `init` is not called. Don't worry about that.

Reply:
Are you sure? It segmentation faults, no error.

Reply:
Damn, that's strange. I looked at the implementation, this api was designed differently (from other apis) -_-. 

For safety concerns, you can add an attached variable (`inited: bool = false`) to the `wrap!` macro(set inited to true after calling `init()`). You can check the `AVSamples` wrapping for how it looks. Then check the `inited` in each function call, if not inited, panic. Or the wrapper is not safe.

While another option is taking advantage of the type System. Create another type: `AVBSFContextUninit`, and all the `AVBSFContext` can only be safely created from it(After calling `av_bsf_init`). This way is cool and fun. :-)